File: packages/python/replica/starfish_replica/http_channel.py
"""``HttpReplicaChannel`` — the original HTTP-pull-into-local-``ObjectStore``
sync mechanics, moved verbatim out of the old ``ReplicaManager._do_sync`` /
``proxy_push``. Mirrors the TS package's ``http-channel.ts``.
"""

from __future__ import annotations

import json
import logging
from typing import Any

import httpx

from starfish_protocol.merge import deep_merge
from starfish_server.protocol.push import push
from starfish_server.router.helpers import deep_sanitize
from starfish_server.protocol.types import PushSuccess
from starfish_server.storage.base import AbstractObjectStore

from starfish_replica.config import RemoteCollection, RemoteConfig, WriteMode
from starfish_replica.channel import ReplicaCallContext

logger = logging.getLogger(__name__)

__all__ = ["HttpReplicaChannel"]


class HttpReplicaChannel:
    """Replicates one :class:`RemoteCollection` from a primary starfish server
    into a local :class:`AbstractObjectStore`."""

    def __init__(
        self,
        store: AbstractObjectStore,
        col: RemoteCollection,
        client: httpx.AsyncClient,
    ) -> None:
        self.name = col.name
        self.remote: RemoteConfig = col.remote
        self._store = store
        self._col = col
        self._client = client
        self._last_hash: str | None = None

    async def sync(self, ctx: ReplicaCallContext) -> None:
        remote = self.remote
        col = self._col

        if remote.write_mode == WriteMode.PUSH_ONLY:
            return

        document_key = col.storage_path

        primary_url = f"{remote.url.rstrip('/')}{remote.pull_path}"
        resp = await self._client.get(
            primary_url,
            headers={"Accept": "application/json", **remote.headers},
        )
        resp.raise_for_status()
        pulled: dict[str, Any] = resp.json()

        primary_hash: str = pulled.get("hash", "")
        primary_data: dict[str, Any] = pulled.get("data", {})

        if not primary_hash:
            return

        if self._last_hash == primary_hash:
            return

        raw_local = await self._store.get_string(document_key)
        current_local_hash: str = ""
        current_local_data: dict[str, Any] = {}
        if raw_local:
            try:
                local_doc = json.loads(raw_local)
                current_local_hash = local_doc.get("hash", "")
                current_local_data = local_doc.get("data", {})
            except json.JSONDecodeError as exc:
                logger.error(
                    "[ReplicaManager] Corrupt local document at %r — treating as empty: %s",
                    document_key, exc,
                )
                # current_local_hash stays "" — push with baseHash="" will overwrite

        if current_local_hash == primary_hash:
            self._last_hash = primary_hash
            return

        if remote.write_mode == WriteMode.BIDIRECTIONAL and current_local_data:
            data_to_write = deep_merge(current_local_data, primary_data)
        else:
            data_to_write = primary_data

        # Strip prototype-pollution keys before writing primary data into the
        # local store. The bidirectional merge drops them via deep_merge, but the
        # pull-only / push-through path writes the primary's ``data`` verbatim and
        # must not trust it — a compromised primary could otherwise plant a
        # ``__proto__`` / ``__class__`` payload.
        sanitized = deep_sanitize(data_to_write)

        # Use current_local_hash directly ("" works for both "no document" and
        # "corrupt document"): push() treats base_hash="" the same as no hash when
        # the stored current_hash is also "". Must NOT coerce "" → None — push()
        # rejects base_hash=None when a (corrupt) doc is present, which would leave
        # a corrupt local doc permanently unrecoverable (sync would raise
        # "Concurrent write" forever). A valid local doc still yields its real
        # hash, so genuine concurrent-write detection is preserved.
        base_hash = current_local_hash
        result = await push(self._store, document_key, sanitized, base_hash)

        if not isinstance(result, PushSuccess):
            raise RuntimeError(
                f"[ReplicaManager] Concurrent write on {self.name!r} — will retry"
            )

        self._last_hash = result.hash
        logger.debug("[ReplicaManager] Synced %r (hash=%s)", self.name, result.hash)

    async def proxy_push(
        self,
        raw_body: bytes | str,
        *,
        on_success: Any = None,
    ) -> tuple[int, Any]:
        """Forward a client push to the primary (write_mode ``push_through``).

        Returns ``(status, body)`` to relay to the client. On success, calls
        ``on_success()`` (if given) so the caller can trigger a background
        sync — this channel does not know about the scheduler that owns it.
        """
        remote = self.remote
        primary_url = f"{remote.url.rstrip('/')}{remote.push_path}"
        headers = {
            "Content-Type": "application/json",
            "Accept": "application/json",
            **remote.headers,
        }

        try:
            resp = await self._client.post(primary_url, content=raw_body, headers=headers)
        except httpx.HTTPError as exc:
            logger.error("Failed to reach primary for %r: %s", self.name, exc)
            return 502, {"error": "Failed to reach primary"}

        if resp.status_code == 409:
            return 409, {"error": "hash_mismatch"}
        if not resp.is_success:
            return resp.status_code, {"error": f"Primary returned {resp.status_code}"}

        body = resp.json()

        # Validate the primary's response shape before relaying it to our client.
        # A successful push returns ``{ hash, timestamp }``; refuse to forward an
        # arbitrary/garbage body a compromised or misbehaving primary might send.
        if not isinstance(body, dict) or not isinstance(body.get("hash"), str):
            logger.error("Primary returned an unexpected push response shape for %r", self.name)
            return 502, {"error": "Primary returned an unexpected response"}

        if on_success is not None:
            on_success()

        return resp.status_code, body

File: packages/ts/replica/src/http-channel.ts
/**
 * The default `ReplicaChannel`: primary→replica-server HTTP pull, reconciled
 * against a local `starfish-server` `ObjectStore`.
 *
 * This is `ReplicaManager`'s original (pre-generalization) `_doSync`/`proxyPush`
 * bodies, moved here VERBATIM — the reconciliation logic (hash short-circuit,
 * `bidirectional` deepMerge, prototype-pollution stripping, corrupt-doc
 * recovery) is unchanged. See `ReplicaManager`'s constructor for how this
 * channel is built from a `RemoteCollection`.
 */
import type { ObjectStore, PushSuccess } from "@drakkar.software/starfish-server"
import { push, deepSanitize } from "@drakkar.software/starfish-server"
import { deepMerge } from "@drakkar.software/starfish-protocol"
import type { RemoteCollection, RemoteConfig } from "./config.js"
import type { ReplicaChannel, ReplicaCallContext } from "./channel.js"

export class HttpReplicaChannel implements ReplicaChannel {
  readonly name: string
  readonly remote: RemoteConfig
  private _store: ObjectStore
  private _storagePath: string
  private _fetch: typeof fetch
  private _lastHash: string | undefined

  constructor(store: ObjectStore, col: RemoteCollection, fetchFn: typeof fetch) {
    this.name = col.name
    this.remote = col.remote
    this._store = store
    this._storagePath = col.storagePath
    this._fetch = fetchFn
  }

  async sync(_ctx: ReplicaCallContext): Promise<void> {
    const remote = this.remote

    if (remote.writeMode === "push_only") return

    const documentKey = this._storagePath
    const primaryUrl = `${remote.url.replace(/\/+$/, "")}${remote.pullPath}`

    const resp = await this._fetch(primaryUrl, {
      headers: { Accept: "application/json", ...remote.headers },
    })
    if (!resp.ok) {
      throw new Error(`Primary returned ${resp.status}`)
    }
    const pulled = (await resp.json()) as Record<string, unknown>

    const primaryHash = (pulled["hash"] as string) ?? ""
    const primaryData = (pulled["data"] as Record<string, unknown>) ?? {}

    if (!primaryHash) return

    if (this._lastHash === primaryHash) return

    const rawLocal = await this._store.getString(documentKey)
    let currentLocalHash = ""
    let currentLocalData: Record<string, unknown> = {}
    if (rawLocal) {
      try {
        const localDoc = JSON.parse(rawLocal) as Record<string, unknown>
        currentLocalHash = (localDoc["hash"] as string) ?? ""
        currentLocalData = (localDoc["data"] as Record<string, unknown>) ?? {}
      } catch (e) {
        console.error(`[ReplicaManager] Corrupt local document at "${documentKey}" — treating as empty:`, e)
        // currentLocalHash stays "" — push with baseHash="" matches the "" stored by push.ts on corrupt read
      }
    }

    if (currentLocalHash === primaryHash) {
      this._lastHash = primaryHash
      return
    }

    let dataToWrite: Record<string, unknown>
    if (
      remote.writeMode === "bidirectional" &&
      Object.keys(currentLocalData).length > 0
    ) {
      dataToWrite = deepMerge(currentLocalData, primaryData)
    } else {
      dataToWrite = primaryData
    }

    // Strip prototype-pollution keys before writing primary data into the local
    // store. The bidirectional merge above already drops them via deepMerge, but
    // the pull-only / push-through path writes the primary's `data` verbatim and
    // must not trust it — a compromised primary could otherwise plant a
    // `__proto__` / `constructor` payload.
    const sanitized = deepSanitize(dataToWrite) as Record<string, unknown>

    // Use currentLocalHash directly ("" works for both "no document" and "corrupt
    // document"): push() treats baseHash="" the same as no hash when the stored
    // currentHash is also "". Must NOT coerce "" → null — push() rejects
    // baseHash=null when a (corrupt) doc is present, which would leave a corrupt
    // local doc permanently unrecoverable (sync would throw "Concurrent write"
    // forever). A valid local doc still yields its real hash, so genuine
    // concurrent-write detection is preserved.
    const baseHash: string = currentLocalHash
    const result = await push(this._store, documentKey, sanitized, baseHash)

    if (!("hash" in result)) {
      throw new Error(
        `[ReplicaManager] Concurrent write on "${this.name}" — will retry`,
      )
    }

    const success = result as PushSuccess
    this._lastHash = success.hash
  }

  /**
   * Forward a client push to the primary (write_mode `push_through`). Returns
   * the response status + body to relay to the client. `onSuccess` is called
   * (not awaited) right before returning a successful response — the manager
   * uses it to trigger a background `syncNow` so `_lastSyncAt`/cooldown state
   * stays owned by the manager, not this channel. Framework-neutral — the
   * caller (replica plugin) turns this into an HTTP response.
   */
  async proxyPush(
    rawBody: string,
    onSuccess?: () => void,
  ): Promise<{ status: number; body: unknown }> {
    const remote = this.remote
    const primaryUrl = `${remote.url.replace(/\/+$/, "")}${remote.pushPath}`
    const headers: Record<string, string> = {
      "Content-Type": "application/json",
      Accept: "application/json",
      ...remote.headers,
    }

    try {
      const resp = await this._fetch(primaryUrl, {
        method: "POST",
        body: rawBody,
        headers,
      })

      if (resp.status === 409) {
        return { status: 409, body: { error: "hash_mismatch" } }
      }
      if (!resp.ok) {
        return { status: resp.status, body: { error: `Primary returned ${resp.status}` } }
      }

      const body = (await resp.json()) as Record<string, unknown>

      // Validate the primary's response shape before relaying it to our client.
      // A successful push returns `{ hash, timestamp }`; refuse to forward an
      // arbitrary/garbage body a compromised or misbehaving primary might send.
      if (
        typeof body !== "object" ||
        body === null ||
        Array.isArray(body) ||
        typeof body["hash"] !== "string"
      ) {
        console.error(`[Starfish] Primary returned an unexpected push response shape for "${this.name}"`)
        return { status: 502, body: { error: "Primary returned an unexpected response" } }
      }

      onSuccess?.()

      return { status: resp.status, body }
    } catch (e) {
      console.error(`[Starfish] Failed to reach primary for "${this.name}":`, e)
      return { status: 502, body: { error: "Failed to reach primary" } }
    }
  }
}
